nncf/experimental/onnx/graph/onnx_graph.py

```diff
@@ -202,24 +205,83 @@ def get_weight_tensor(self, node: onnx.NodeProto) -> Tuple[str, np.ndarray]:
         :param node: Node, in which the weight tensor finally applied.
         :return: Weight tensor name and its value.
         """
-        node_name = node.name
-        node_inputs = self.get_node_edge_names(node_name)['input']
-        weight_input = node_inputs[1]
         try:
+            weight_port_id = self.get_weight_tensor_port_id(node)
+            weight_input = self.get_node_edge_names(node.name)['input'][weight_port_id]
             return self.get_initializer(weight_input).name, self.get_initializers_value(weight_input)
-        except RuntimeError as e:
-            node = self.get_nodes_by_output(weight_input)[0]
-            metatype = ONNX_OPERATION_METATYPES.get_operator_metatype_by_op_name(node.op_type)
-            if metatype == ONNXIdentityMetatype:
-                tensor_name = self.get_initializer(node.input[0]).name
-                return tensor_name, self.get_initializers_value(tensor_name)
-            if metatype == ONNXReshapeMetatype:
-                tensor_name = node.output[0]
-                shape = self.get_initializers_value(node.input[1])
-                tensor_value = self.get_initializers_value(node.input[0])
-                reshaped_tensor_value = tensor_value.reshape(shape)
-                return tensor_name, reshaped_tensor_value
-            raise RuntimeError('Could not find the weight value of the node') from e
+        except RuntimeError as exc:
+            weight_port_id = self.get_weight_tensor_port_id(node)
+            parent_node_on_weight_port = self.get_parents(node)[weight_port_id]
+            nodes = deque([parent_node_on_weight_port])
+            while nodes:
+                try:
```

Comment:
Two levels of `try-except`? Really?

Reply:
I tend to agree that the code is convoluted and that we can miss real RuntimeErrors in here. It was OK when the try-except block was limited to a single function, but no longer. Refactor the `get_` function to not throw exceptions as part of their normal functionality, and set up `def has_*(name) -> bool:` functions instead for each function that used to throw exceptions. Call the `has_*` function in an if-block before calling the `get_` functions and do early-exit if `has_` check failed.

Also, this much logic has to be unit-tested.

Reply:
Done